yarn-project/end-to-end/src/e2e_2_rpc_servers.test.ts

```diff
@@ -0,0 +1,148 @@
+import { ZkTokenContract } from '@aztec/noir-contracts/types';
+import { L2BlockL2Logs, LogType, TxStatus } from '@aztec/types';
+
+import { setup, setupWithoutDeployingContractsAndAztecNode } from './utils.js';
+
+describe('e2e_2_rpc_servers', () => {
+  let aztecNode: AztecNodeService;
+  let aztecRpcServerA: AztecRPCServer;
+  let aztecRpcServerB: AztecRPCServer;
+  let walletA: Wallet;
+  let walletB: Wallet;
+  let userA: AztecAddress;
+  let userB: AztecAddress;
+  let logger: DebugLogger;
+
+  let contractWithWalletA: ZkTokenContract;
+  let contractWithWalletB: ZkTokenContract;
+
+  beforeEach(async () => {
+    let accounts: AztecAddress[] = [];
+    ({ aztecNode, aztecRpcServer: aztecRpcServerA, accounts, wallet: walletA, logger } = await setup(1));
+    [userA] = accounts;
+
+    ({
+      aztecRpcServer: aztecRpcServerB,
+      accounts: accounts,
+      wallet: walletB,
+    } = await setupWithoutDeployingContractsAndAztecNode(1, aztecNode));
+    [userB] = accounts;
+  }, 100_000);
+
+  afterEach(async () => {
+    await aztecNode?.stop();
+    await aztecRpcServerA?.stop();
+    await aztecRpcServerB?.stop();
+  });
+
+  const expectBalance = async (wallet: Wallet, owner: AztecAddress, expectedBalance: bigint) => {
+    const contractWithWallet = new ZkTokenContract(contractWithWalletA.address, wallet);
+    const [balance] = await contractWithWallet.methods.getBalance(owner).view({ from: owner });
+    logger(`Account ${owner} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  const expectsNumOfEncryptedLogsInTheLastBlockToBe = async (numEncryptedLogs: number) => {
+    const l2BlockNum = await aztecNode.getBlockHeight();
+    const encryptedLogs = await aztecNode.getLogs(l2BlockNum, 1, LogType.ENCRYPTED);
+    const unrolledLogs = L2BlockL2Logs.unrollLogs(encryptedLogs);
+    expect(unrolledLogs.length).toBe(numEncryptedLogs);
+  };
+
+  const expectUnencryptedLogsFromLastBlockToBe = async (logMessages: string[]) => {
+    const l2BlockNum = await aztecNode.getBlockHeight();
+    const unencryptedLogs = await aztecNode.getLogs(l2BlockNum, 1, LogType.UNENCRYPTED);
+    const unrolledLogs = L2BlockL2Logs.unrollLogs(unencryptedLogs);
+    const asciiLogs = unrolledLogs.map(log => log.toString('ascii'));
+
+    expect(asciiLogs).toStrictEqual(logMessages);
+  };
+
+  const deployContract = async (initialBalance: bigint, from: AztecAddress) => {
+    logger(`Deploying L2 contract...`);
+    const tx = ZkTokenContract.deploy(aztecRpcServerA, initialBalance, from).send();
+    const receipt = await tx.getReceipt();
+    contractWithWalletA = new ZkTokenContract(receipt.contractAddress!, walletA);
+    contractWithWalletB = new ZkTokenContract(receipt.contractAddress!, walletB);
+    await tx.isMined(0, 0.1);
+    const minedReceipt = await tx.getReceipt();
+    expect(minedReceipt.status).toEqual(TxStatus.MINED);
+    logger('L2 contract deployed');
+    return contractWithWalletA;
+  };
```

Comment:
I'd move this to a `beforeEach` directly, since it's setting up test variables. But it's a nitpick.